libyul/backends/evm/SSAControlFlowGraphBuilder.cpp

```diff
@@ -626,22 +615,29 @@ SSACFG::ValueId SSAControlFlowGraphBuilder::readVariableRecursive(Scope::Variabl
 	auto& block = m_graph.block(_block);
 	auto& info = blockInfo(_block);
 
-	if (info.sealed && block.entries.size() == 1)
-		return readVariable(_variable, *block.entries.begin());
-
-	SSACFG::ValueId phi = m_graph.newPhi(_block);
-	block.phis.insert(phi);
-	if (info.sealed)
+	SSACFG::ValueId val;
+	if (!info.sealed)
 	{
-		writeVariable(_variable, _block, phi);
-		phi = addPhiOperands(_variable, phi);
+		// incomplete block
+		val = m_graph.newPhi(_block);
+		block.phis.insert(val);
+		info.incompletePhis.emplace_back(val, std::ref(_variable));
 	}
+	else if (block.entries.size() == 1)
+		// one predecessor: no phi needed
+		val = readVariable(_variable, *block.entries.begin());
 	else
 	{
-		info.incompletePhis.emplace_back(phi, std::ref(_variable));
+		// Break potential cycles with operandless phi
+		val = m_graph.newPhi(_block);
+		block.phis.insert(val);
+		writeVariable(_variable, _block, val);
+		// we call tryRemoveTrivialPhi explicitly opposed to what is presented in Algorithm 2, as our implementation
+		// does not call it in addPhiOperands to avoid removing phis in unsealed blocks
+		val = tryRemoveTrivialPhi(addPhiOperands(_variable, val));
 	}
-	writeVariable(_variable, _block, phi);
-	return phi;
+	writeVariable(_variable, _block, val);
```

Comment:
For the record: the ``info.sealed && block.entries.size() == 1`` case *should* also have done this from the start (I guess that decreases overall complexity) - I was just careless in flipping the conditions from the paper - reproducing the exact structure from the paper as done now is indeed better than trying to be clever and then getting it wrong!

Reply:
Yup, we can still be clever when it's all tested:)